Initial changes for supporting beanshell

krmahadevan opened the pull request.

krmahadevan commented:
@juherr - Please let me know if the interface definition and the pom file looks fine. Those are the only two unique things to this project. We will be injecting the implementation using SPI from within TestNG. The interface in this project will move into TestNG and once that is merged, this project will again be updated to start referring to TestNG `7.0.0-SNAPSHOT`. Just giving you a heads up that many iterations would perhaps be required on this before it comes to a point of stability.

juherr commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+  </dependencies>
+
+    <build>
+        <plugins>
```
I prefer plugin management (https://maven.apache.org/pom.html#Plugin_Management) for plugins version/configuration and plugins for activation.
But here, it is not really relevant because we don't have too much configuration

juherr commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>2.19.1</version>
+            </plugin>
+        </plugins>
```
You can freeze the jar plugin version (and deploy + release) too

krmahadevan replied:
For the deploy and release, I have been following the instructions detailed [here](https://github.com/making/travis-ci-maven-deploy-skelton). I need @cbeust help in getting them done. So for now I am not getting into that section yet.

juherr commented on src/main/java/org/testng/BeanShellScriptedMethodSelector.java:
```diff
@@ -0,0 +1,84 @@
+    return InterpreterHolder.getInstance();
+  }
+
+  private void setContext(
```
static

krmahadevan replied:
fixed

juherr commented on src/main/java/org/testng/BeanShellScriptedMethodSelector.java:
```diff
@@ -0,0 +1,84 @@
+    }
+  }
+
+  private void resetContext(Interpreter interpreter) {
```
static

krmahadevan replied:
fixed

juherr commented on src/main/java/org/testng/IScriptedMethodSelector.java:
```diff
@@ -0,0 +1,28 @@
+   * @return - A String constant that TestNG should use to register the implementation as the
+   *     supported language.
+   */
+  String language();
```
I prefer `boolean accept(String language)` because it will transfer the logic from testng core to implementation (which is the best to know its capabilities).

You can keep language but it may support many languages: `String[] languages()` (or `Set<String>`)

krmahadevan replied:
I guess I will stick to the `boolean accept(String language)` part. It looks elegant and we are resorting to kind of like an inversion of control :)

juherr commented on src/main/java/org/testng/IScriptedMethodSelector.java:
```diff
@@ -0,0 +1,28 @@
+// the sake
+// of ensuring that this project can first compile properly, its being temporarily accomodated here.
+// This interface should move into TestNG 7.0.0 before it gets released.
+public interface IScriptedMethodSelector {
```
I will separate the build logic and the selection logic:

```java
public interface IScriptedMethodSelectorFactory {
  boolean accept(String language);
  IMethodSelector create(String language, String expression); // Maybe TestContext or ISuite or XmlSuite can help here
  Set<String> languages();
}
```

So we will be able to reuse some current interface and logic.

krmahadevan replied:
I wouldn't want to change the signature. I believe it would all boil down to the ScriptBasedMethodSelector returning `true` if the current method is to be included or `false` otherwise. Why do we need to construct a `IMethodSelector` again? Also now that we using a `boolean accept()`, I guess there's no need for the `Set<String> languages()` because originally I was thinking of using an SPI to register all implementations into TestNG as a Key-value pair and then based on the language invoke the corresponding implementation. But now that even that been delegated back to the actual implementation, TestNG doesn't need to know any of the languages information etc.,

juherr replied:
I don't have the global picture. Could point me where in the core you'd like to use the class?

krmahadevan replied:
Here's what I have 
1. We need to enhance `org.testng.internal.XmlMethodSelector` so that it loads the `org.testng.IScriptedMethodSelector` implementations via SPI and within `org.testng.internal.XmlMethodSelector#includeMethod()` it first tests if `org.testng.IScriptedMethodSelector#accept(language)` returns true and if its true, then we invoke `org.testng.IScriptedMethodSelector#includeMethodFromExpression()`.

This is the plan.

juherr replied:
Ok. Mine is more:

XmlMethodSelector and ScriptedMethodSelector are 2 different implementations of IMethodSelector.
Both are created from the XmlSomething

krmahadevan replied:
@juherr I chose this, because this is less invasive and I dont have to do a lot of changes to accommodate this de-coupling. So I hope we can go ahead with this plan.

juherr replied:
I feel some perf issues with your solution because we will have to compile the same expression for each method.
It won't have any impact on the current implementation with BeanShell but I can easily imagine some other script engines where we will be able to compile the expression first and pass the parameters after.

krmahadevan commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.0</version>
+                <configuration>
+                    <source>1.8</source>
```
fixed.

krmahadevan commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+            </plugin>
+        </plugins>
+    </build>
+    <distributionManagement>
```
Who would be doing that sync? Does it happen automatically or should we do it manually ?

juherr replied:
It should/could be automatic.

krmahadevan commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+  </dependency>
+    <dependency>
+      <groupId>org.testng</groupId>
+      <artifactId>TestNG</artifactId>
```
fixed.

krmahadevan commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+    </issueManagement>  
+
+  <dependencies>
+  <dependency>
```
my bad. Fixed.

krmahadevan commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+        </developer>
+    </developers>
+    <issueManagement>
+        <url>https://github.com/testng-team/testng-script-beanshell</url>
```
fixed

krmahadevan commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+                <role>Owner</role>
+            </roles>
+        </developer>
+    </developers>
```
Now that we are talking of adding others, added me and you :)

juherr replied:
👍

krmahadevan commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+        <developer>
+            <id>cbeust</id>
+            <name>Cedric Beust</name>
+            <url>https://github.com/cbeust</url>
```
Fixed

krmahadevan commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
```
my bad. bad formatting. fixed

juherr replied:
I still see the empty line ;)

krmahadevan replied:
fixed.

krmahadevan commented:
@juherr - I have amended the commit and pushed in the fixes. Please take a look

krmahadevan commented:
@juherr - with respect to the BinTray settings for deployment, I could use some help. I am not familiar with that. Do you think you could help by maybe targeting a PR just for that, once this changes are merged? I still have a lot of changes pending.

juherr commented on pom.xml:
```diff
@@ -0,0 +1,112 @@
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>3.0</version>
+        <configuration>
```
In fact you can remove the configuration part. The plugin will use the values from the properties itself

juherr replied:
Ping

krmahadevan commented:
@juherr - I have addressed the review comments (amended the existing commit and pushed the changes). Please check. I have tried addressing the publishing to maven also. We can give this a try and see if that works (We may need @cbeust to update the travis settings page for this repository with the sonatype credentials so that the publishing works)

krmahadevan commented:
ping @juherr - pls let me know if there is anything additional that needs to be taken care of, before getting this merged. I was hoping that I can wrap up this entire feature (by making changes in both this library and testng within this weekend). :)

juherr commented on pom.xml:
```diff
@@ -0,0 +1,176 @@
+        <version>2.19.1</version>
+      </plugin>
+
+      <plugin>
```
Make `release` profile because we only need some plugins for the release.

krmahadevan replied:
@juherr - If we make a `release` profile then we will have to also alter the `.travis.yml` to take care of invoking the maven command with that profile only when doing the release. Currently only javadocs and sources plugin are added to the pom file and I think we should be fine with publishing javadocs and sources even for snapshots as well.  I also dont expect this artifact to go through with many releases. So I guess we should be ok, without having a release profile.

krmahadevan commented:
@juherr 

> I feel some perf issues with your solution because we will have to compile the same expression for each method.
It won't have any impact on the current implementation with BeanShell but I can easily imagine some other script engines where we will be able to compile the expression first and pass the parameters after.

I don't think it would be TestNG that is going to be doing it, but its going to be delegating the call to the provider. So as per our original plan of now wanting to support Javascript based expressions, shouldn't the ability to basically cache the parsed expression now lie with the implementation? TestNG merely is passing in the expression and the `Test` method. The implementation should be worrying about this I guess and not TestNG. So the perf optimization would not be on TestNG

juherr commented:
> to be delegating the call to the provider

Exactly! That's why we need to separate the expression part and the method selection part.
The factory pattern is a way to do that because the implementation will be able to cache/optimize the expression, and then TestNG will use the same expression impl for each method.

krmahadevan commented:
@juherr - with respect to the perf issues you are hinting at, heres the flow

1. TestNG first dynamically loads all implementations of `org.testng.IScriptedMethodSelector` via SPI.
2. When TestNG is confronted with an expression, TestNG runs through all the implementations of `org.testng.IScriptedMethodSelector` that it is aware of until it finds the first implementation that returns `true` for `accept()` method. Now for this implementation TestNG passes the expression and the `ITestNGMethod` object and expects the implementation to return `true` or `false`. So it would be the responsibility of `org.testng.IScriptedMethodSelector` implementation (for e.g., `org.testng.BeanShellScriptedMethodSelector` to take care of caching if required etc., )

This is what I had in mind. Can you please tell me where in this do you see perf issues?

Its the implementation specifics that will have to take care of how it wants to basically parse the expression (should it parse it every time or parse if not already parsed or use the parsed evaluation if its the same expression) and work with it, and TestNG shouldn't be concerned.

krmahadevan commented:
> What is the issue with a factory pattern?

I didnt quite understand the application of the factory pattern especially
when its the implementations that are going to be providing this to TestNG
(something like a dependency injection from TestNG perspective), wherein
TestNG would delegate the call to the downstream implementations. How does
the factory pattern fit there. Any sample that you could please share to
help me understand ?

> How will you manage if you need a new interpreter by expression?

By caching the expressions with their interpreters via a Map maybe ? [
Map<String, Interpreter> ]

> How do you free the memory in case of a cache of interpreters?

The implementation providers of our interface could easily handle this with
a WeakReference map no ? since the interpreter instantiation is not backed
by any state, and is purely "create one more instance if the existing
instance in the map is null",  I thought it should be fairly straight
forward. Am I missing something here ?

juherr commented:
> The implementation providers of our interface could easily handle this with a WeakReference map no ? since the interpreter instantiation is not backed by any state, and is purely "create one more instance if the existing instance in the map is null", I thought it should be fairly straight forward. Am I missing something here ?

That's the point I'd like to avoid because I think interpreters are high consumers of time and memory.

juherr commented:
In fact, I don't see it as an optimization because it is just using the GC cleaning the memory when we will have, "perhaps", to find some workaround later with the current implementation.
The point is that it doesn't cost to add (or reuse) an interface in order to separate the creation logic from the usage logic.
The current implementation is not wrong but has 2 different purposes.

krmahadevan commented:
> The point is that it doesn't cost to add (or reuse) an interface in order
to separate the creation logic from the usage logic.

I think I am going around in circles and perhaps not understanding what you
are saying. Would it be possible for you to jot down some sample code for
me, so that it makes your idea clear? I am still not able to grasp as to
where does a factory pattern (that you mentioned earlier as well) fit in
here and what changes are you suggesting to the current implementation.

juherr commented:
No prob, I will try it asap

cbeust commented:
Isn't all of this (managing interpreter references, matching them to languages, etc...) automatically handled by the native scripting support in the JDK?

krmahadevan commented:
@cbeust - By native API are you referring to `javax.script.ScriptEngine` interface and its implementations ?

If yes, then beanshell  (`bsh.engine.BshScriptEngine`) also implements this interface .

cbeust commented:
Yes, I was referring to `ScriptEngine`, which discovers available languages through service loaders (whatever implementation of that interface that happens to be on the classpath is made available).

krmahadevan commented:
@cbeust - Thanks for clarifying that.

@juherr 

> drop beanshell feature

I believe this entire effort of extracting out the functionality of beanshell was to help retain the feature. Its a lot more easier to write script to filter methods than to write java code 😋 

> bind beanshell to the jdk api

I believe that the beanshell implementation is already doing that (I looked at the implementation and also confirmed the doubt with @cbeust )

> provide a custom api and 2 impl (beanshell + jdk api)

Not sure what you mean by this.  The JDK API is the only implementation that exists for Beanshell as of today. Its just that it has been sugar coated and abstracted out under multiple layers to make it a one line invocation.  But I would like to know more on what you are hinting at.

I also am waiting for the sample that you said you would share. This PR has been pending for a few days now and I was hoping we could wrap this to closure before moving on to the next one.

juherr commented:
@krmahadevan Finally, I found enough time to have a look at this subject: https://github.com/cbeust/testng/pull/1847
Everything is not yet good enough but it is a good step IMO.

I think we can delete this repository once the pull request will be merged.

krmahadevan commented:
@juherr - Thanks for raising that PR. That clarifies your intent. Yes we can now delete off this repository.  I think I will now close off this PR
